sys/src/9/amd64/dat.h

```diff
@@ -354,69 +352,88 @@ struct Stackframe
 };
 
 /*
- * This is the low memory map, between 0x100000 and 0x110000.
- * It is located there to allow fundamental datastructures to be
+ * This is the low memory map, between 1MiB and 2MiB.
+ *
+ * It is located there to allow fundamental data structures to be
  * created and used before knowing where free memory begins
  * (e.g. there may be modules located after the kernel BSS end).
- * The layout is known in the bootstrap code in l32p.s.
+ * The layout is known in the bootstrap code in entry.S
+ *
  * It is logically two parts: the per processor data structures
  * for the bootstrap processor (stack, Mach, vsvm, and page tables),
  * and the global information about the system (syspage, ptrpage).
  * Some of the elements must be aligned on page boundaries, hence
  * the unions.
  */
 struct Sys {
-	unsigned char	machstk[MACHSTKSZ];
+	alignas(4096) unsigned char machstk[MACHSTKSZ];
 
-	PTE	pml4[PTSZ/sizeof(PTE)];	/*  */
-	PTE	pdp[PTSZ/sizeof(PTE)];
-	PTE	pd[PTSZ/sizeof(PTE)];
-	PTE	pt[PTSZ/sizeof(PTE)];
+	PTE	ipml4[PTSZ/sizeof(PTE)];	// Only used very early in boot
```

Comment:
ipml4 is the initial PML4 that contains the double map of the first 4GiB of physical address space: one mapping is an identity map, the other is a direct offset map at KZERO.

epml4 is the same as ipml4 but with the identity mapping removed; we burn a page on that, but don't have to coordinate "zapping" the identity map with the APs. The BSP transitions to running on epml4 in warp64. pml4 is the actual PML4 that the system lives on: this properly maps everything in the physical address space, including making sure that kernel text is RO, everything except text is NXE, and .rodata is RO. It holds mappings for PCI BARs and stuff. But we can't set most of that up until we get into C code (or, rather, it would be really tedious to set it up in assembler), so we enter the kernel on the BSP on epml4, then build up pml4 and then transition to it in `meminit`. The APs start on ipml4 (which retains the identity mapping), but move to pml4 before jumping to squidboy.

Reply:
thanks - would be good to have this comment in the code, if it's not elsewhere